File: packages/research_management/src/research_management/models/member.py
"""
Project member models.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class MemberRole(str, Enum):
    """Member roles in a research project."""

    STUDENT = "student"
    ADVISOR = "advisor"
    CO_ADVISOR = "co-advisor"
    COORDINATOR = "coordinator"


class ProjectMember(BaseModel):
    """Project member model."""

    project_id: str = Field(..., description="Project identifier")
    user_id: str = Field(..., description="User identifier")
    role: MemberRole = Field(..., description="Member role in the project")
    joined_at: datetime = Field(
        default_factory=datetime.utcnow, description="When member joined"
    )
    left_at: Optional[datetime] = Field(None, description="When member left (if any)")

    class Config:
        json_schema_extra = {
            "example": {
                "project_id": "proj-123",
                "user_id": "user-456",
                "role": "student",
                "joined_at": "2025-01-15T10:00:00Z",
                "left_at": None,
            }
        }


class ProjectMemberCreate(BaseModel):
    """Model for adding a member to a project."""

    user_id: str = Field(..., description="User identifier")
    role: MemberRole = Field(..., description="Member role in the project")

File: packages_dashboard/research_dashboard/lib/src/models/member.dart
/// Project member model
class ProjectMember {
  final String projectId;
  final String userId;
  final MemberRole role;
  final DateTime joinedAt;
  final DateTime? leftAt;

  const ProjectMember({
    required this.projectId,
    required this.userId,
    required this.role,
    required this.joinedAt,
    this.leftAt,
  });

  factory ProjectMember.fromJson(Map<String, dynamic> json) {
    return ProjectMember(
      projectId: json['project_id'] as String,
      userId: json['user_id'] as String,
      role: MemberRole.fromString(json['role'] as String),
      joinedAt: DateTime.parse(json['joined_at'] as String),
      leftAt: json['left_at'] != null
          ? DateTime.parse(json['left_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'project_id': projectId,
      'user_id': userId,
      'role': role.value,
      'joined_at': joinedAt.toIso8601String(),
      'left_at': leftAt?.toIso8601String(),
    };
  }
}

/// Member role enum
enum MemberRole {
  student('student'),
  advisor('advisor'),
  coAdvisor('co-advisor'),
  coordinator('coordinator');

  final String value;
  const MemberRole(this.value);

  static MemberRole fromString(String value) {
    return MemberRole.values.firstWhere((e) => e.value == value);
  }

  String get displayName {
    switch (this) {
      case MemberRole.student:
        return 'Aluno';
      case MemberRole.advisor:
        return 'Orientador';
      case MemberRole.coAdvisor:
        return 'Co-orientador';
      case MemberRole.coordinator:
        return 'Coordenador';
    }
  }
}
